File: Mobile/ios/swift/Sys.swift
//
//  Sys.swift
//  SampleSwiftUI
//
//  Created by Karim Lalani on 1/17/20.
//  Copyright © 2020 Sertainty. All rights reserved.
//

import Foundation

public class Sys {
    public class func getAppRoot() -> String? {
        return uxpsys_getAppRoot().map {
            (ptr: UnsafePointer<Int8>) -> String in
            String(cString: ptr)
        }
    }
    
    public class func setAppRoot(_ path :String) {
        uxpsys_setAppRoot(path)
    }
    
    public class func initLibrary(baError: ByteArray,
                           appKey: String,
                           prefix: String,
                           version: String,
                           onInit: @escaping ()->Void,
                           onFail: @escaping ()->Void) {
        let argv = UnsafeMutablePointer<UnsafeMutablePointer<Int8>?>.allocate(capacity: 1)
        let argv1 = UnsafeMutablePointer<Int8>(mutating: appKey)
        argv.assign(repeating: argv1, count: 1)
        
        let dispatchQueue = DispatchQueue(label: "InitializeLibrary")
        dispatchQueue.async {
            if uxpsys_initLibrary(baError.handle, 1, argv, "sertainty.lic", "SertintyONE", prefix, version) == 1 {
                DispatchQueue.main.async {
                    onInit()
                }
            } else {
                DispatchQueue.main.async {
                    onFail()
                }
            }
        }
    }
    
    public class func installLicense(baError: ByteArray,
                                     licenseFile: String,
                                     onInstall: @escaping ()->Void,
                                     onFail: @escaping () -> Void)
    {
        DispatchQueue(label: "InstallLicense").async {
            if uxpsys_installLicense(baError.handle, licenseFile) == 1 {
                DispatchQueue.main.async {
                    onInstall()
                }
            } else {
                DispatchQueue.main.async {
                    onFail()
                }
            }
        }
    }
    
    public class func fileReadAll(_ callStatus: CallStatus, _ fileSpec: String, _ buffer: ByteArray)
    {
        uxpsys_fileReadAll(callStatus.handle,fileSpec,buffer.handle)
    }
    
    public class func fileWriteAll(_ callStatus: CallStatus, _ fileSpec: String, _ data: Data, _ mods: Int)
    {
        let array = data.map{
            (item: UInt8) -> Int8 in
            Int8(item)
        }
        uxpsys_fileWriteAll(callStatus.handle,fileSpec,array,uxp_int64( data.count),Int32(mods))
    }
}

File: Mobile/android/wrappers/java/org/Sertainty/Sys.java
package org.Sertainty;

public class Sys {
    static {
        System.loadLibrary(Library.SertaintyCore());
    }

    public static class CallStatus{
        public CallStatus(){
            handle = newCallStatusHandle();
        }

        public CallStatus(long handle)
        {
            this.handle = handle;
            this.derived = true;
        }

        public boolean hasError()
        {
            return Sys.hasError(handle);
        }

        public String getErrorMessage()
        {
            return Sys.getErrorMessage(handle);
        }

        public long getHandle()
        {
            return handle;
        }



        @Override
        protected void finalize() throws Throwable {
            super.finalize();
            if(!derived) {
                freeCallStatusHandle(handle);
            }
        }

        private long handle;
        private boolean derived = false;
    }

    public static boolean initLibrary(ByteArray errBuffer, String appName, String prefix, String version)
    {
        return initLibrary(errBuffer.getHandle(),appName,prefix,version);
    }

    public static boolean installLicense(ByteArray errBuffer, String licenseFile)
    {
        return installLicense(errBuffer.getHandle(),licenseFile);
    }

    public static boolean releaseLibrary(ByteArray errBuffer)
    {
        return releaseLibrary(errBuffer.getHandle());
    }

    public static void fileReadAll(CallStatus callStatus, String fileSpec, ByteArray outbuf)
    {
        fileReadAll(callStatus.getHandle(),fileSpec,outbuf.getHandle());
    }

    public static void fileWriteAll(CallStatus callStatus, String fileSpec, byte[] data, int mods)
    {
        fileWriteAll(callStatus.getHandle(),fileSpec,data,mods);
    }

    private native static long newCallStatusHandle();
    private native static void freeCallStatusHandle(long handle);
    public native static void setLogging(int level);
    public native static void setLogFile(String prefix, String version);
    private native static boolean initLibrary(long baerror, String appName, String prefix, String version);
    private native static boolean releaseLibrary(long baerror);
    private native static void fileReadAll(long callStatus, String fileSpec, long outbuf);
    private native static void fileWriteAll(long callStatus, String fileSpec, byte[] data, int mods);
    private native static boolean hasError(long handle);
    private native static String getErrorMessage(long handle);
    private native static boolean installLicense(long baerror, String licenseFile);
}
